File: leetcode/leetcode_cpp/construct-quad-tree.cpp
/*
// Definition for a QuadTree node.
class Node {
public:
    bool val;
    bool isLeaf;
    Node* topLeft;
    Node* topRight;
    Node* bottomLeft;
    Node* bottomRight;
    
    Node() {
        val = false;
        isLeaf = false;
        topLeft = NULL;
        topRight = NULL;
        bottomLeft = NULL;
        bottomRight = NULL;
    }
    
    Node(bool _val, bool _isLeaf) {
        val = _val;
        isLeaf = _isLeaf;
        topLeft = NULL;
        topRight = NULL;
        bottomLeft = NULL;
        bottomRight = NULL;
    }
    
    Node(bool _val, bool _isLeaf, Node* _topLeft, Node* _topRight, Node* _bottomLeft, Node* _bottomRight) {
        val = _val;
        isLeaf = _isLeaf;
        topLeft = _topLeft;
        topRight = _topRight;
        bottomLeft = _bottomLeft;
        bottomRight = _bottomRight;
    }
};
*/

/*
https://leetcode.com/problems/construct-quad-tree

Solution1. Divide and Conquer (recursion)
time: o(n^2logn)
space: o(logn)
*/

class Solution {
public:
    bool isAllSame(vector<vector<int>>& grid, int x, int y, int n) {
        int val = grid[x][y];
        for (int i=x; i<x+n; ++i) {
            for (int j=y; j<y+n; ++j) {
                if (val != grid[i][j])
                    return false;
            }
        }
        return true;
    }
    Node* build(vector<vector<int>>& grid, int x, int y, int n) {
        if (isAllSame(grid, x, y, n))
            return new Node(grid[x][y], true);

        n /= 2;
        Node* node = new Node(true, false);
        node->topLeft = build(grid, x, y, n);
        node->topRight = build(grid, x, y+n, n);
        node->bottomLeft = build(grid, x+n, y, n);
        node->bottomRight = build(grid, x+n, y+n, n);
        return node;
    }
    Node* construct(vector<vector<int>>& grid) {
        return build(grid,0,0,grid.size());
    }
};

File: leetcode/leetcode_py/construct-quad-tree.py
"""
# Definition for a QuadTree node.
class Node:
    def __init__(self, val, isLeaf, topLeft, topRight, bottomLeft, bottomRight):
        self.val = val
        self.isLeaf = isLeaf
        self.topLeft = topLeft
        self.topRight = topRight
        self.bottomLeft = bottomLeft
        self.bottomRight = bottomRight
"""

'''
https://leetcode.com/problems/construct-quad-tree

Solution1. Divide and Conquer (recursion)
time: o(n^2logn)
space: o(logn)
'''

class Solution:
    def construct(self, grid: List[List[int]]) -> 'Node':
        def isAllSame(x, y, n):
            val = grid[x][y]
            for i in range(x, x+n):
                for j in range(y, y+n):
                    if val != grid[i][j]:
                        return False
            return True
        def build(x, y, n):
            if isAllSame(x, y, n):
                return Node(grid[x][y], True)
            n //= 2
            node = Node(True, False)
            node.topLeft = build(x, y, n)
            node.topRight = build(x, y+n, n)
            node.bottomLeft = build(x+n, y, n)
            node.bottomRight = build(x+n, y+n, n)
            return node
        return build(0, 0, len(grid))

        